### PR TITLE
protoc-gen-openapiv2: RPC visibility setting transitively applied to messages

### PR DESCRIPTION
**Pull request by erademacher**
#### References to other Issues or PRs

Fixes [grpc-ecosystem#2691](https://github.com/grpc-ecosystem/grpc-gateway/issues/2691)

#### Have you read the [Contributing Guidelines](https://github.com/grpc-ecosystem/grpc-gateway/blob/master/CONTRIBUTING.md)?

Yes!

#### Brief description of what is fixed or changed

Skip hidden methods in findServicesMessagesAndEnumerations, transitively hiding their requests and responses unless they're included in another visible method.

#### Other comments

Not sure what happened with the double commits. I'll be sure to squash them.

### CODE REVIEW

**Review by johanbrandhorst — COMMENTED** [review]
Thank you for your contribution! I'm wondering if this is deceptively complicated. We have to consider what happens when merging different packages into one openapi file. If we just automatically skip messages because an RPC using that message is hidden, we could end up with unsatisfied references for RPCs that use imported messages. Here's a quote from the PR that originally added this (https://github.com/grpc-ecosystem/grpc-gateway/pull/2578): 

> Support is also missing for adding annotations to messages and enums, this was excluded due to complexity and potential edge cases when excluding a type that is required somewhere else in the OpenAPI Schema.

I wonder if this is going to be possible at all without some sort of meta-analysis of the final, _merged_ file. CC @BCook98.

**Comment by BCook98** [comment]
This seems like a good solution to the issue. Could do with the extra test case I suggest around sub messages.

> We have to consider what happens when merging different packages into one openapi file. If we just automatically skip messages because an RPC using that message is hidden, we could end up with unsatisfied references for RPCs that use imported messages. 

In that case it should either fail because the relevant `.proto` files weren't imported or included. 
https://github.com/grpc-ecosystem/grpc-gateway/blob/672f079d1ec726166bbdf413b1fa24bf6da2cb35/internal/descriptor/registry.go#L285

Or if the message is in the registry because it was correctly imported/included, then at the point in time it is referenced by a method that is visible it will be rendered. As it should always push messages into message map as they are referenced.
https://github.com/grpc-ecosystem/grpc-gateway/blob/b62dd44331c17bb28658371f8cc9ccdb1db9bfbe/protoc-gen-openapiv2/internal/genopenapi/template.go#L377, rendering: https://github.com/grpc-ecosystem/grpc-gateway/blob/b62dd44331c17bb28658371f8cc9ccdb1db9bfbe/protoc-gen-openapiv2/internal/genopenapi/template.go#L578

My concerns were around specifically having the ability to manually decided if a message/enum should be rendered, as a that makes it far easier to accidentally create invalid references.

File: examples/internal/proto/examplepb/visibility_rule_echo_service.proto [hunk 1]
```diff
@@ -53,6 +53,13 @@ message VisibilityRuleSimpleMessage {
   VisibilityEnum an_enum = 10;
 }
 
+// MessageInHiddenMethod doesn't define its own visibility restrictions,
+// but is only included in a method marked as "PREVIEW", so it will only
+// appear if `visibility_restriction_selectors` include "PREVIEW".
+message VisibilityRuleMessageInPreviewMethod {
+  string id = 1;
+}
+
```
[the remark this answers is not in the document]
Comment (erademacher) [reply examples/internal/proto/examplepb/visibility_rule_echo_service.proto]: Added a submessage and an enum to make sure they're both hidden.

**Comment by johanbrandhorst** [comment]
@erademacher have you signed the CLA? Sometimes the bot is a bit buggy and doesn't update, in which case I can override it, but please share a screenshot of your CLA status if you think the bot is wrong.

**Comment by erademacher** [comment]
Sorry for the delay. Had to check with legal to make sure I was ok to sign the CLA.

**Comment by johanbrandhorst** [comment]
Looks like there's an issue with the naming of one of the new enum values, can you take a look?
```
$ buf lint
examples/internal/proto/examplepb/visibility_rule_echo_service.proto:76:4:Enum value name "VISIBILITY_PREVIEW_ENUM_UNSPECIFIED" should be prefixed with "VISIBILITY_RULE_ENUM_IN_PREVIEW_METHOD_".
```

**Comment by johanbrandhorst** [comment]
Looks like the linter problem is still an issue

**Comment by erademacher** [comment]
I'm used to a different git workflow and I managed to majorly mess up my fork, so I'm starting over. New PR is #2880.

**Pull request closed without merging**